https://github.com/ToluClassics/candle-tutorial

In [12]:
%cd /content

/content


In [13]:
!git clone https://github.com/huggingface/candle.git

Cloning into 'candle'...
remote: Enumerating objects: 22268, done.
remote: Counting objects: 100% (291/291), done.
remote: Compressing objects: 100% (189/189), done.
remote: Total 22268 (delta 211), reused 102 (delta 102), pack-reused 21977 (from 3)
Receiving objects: 100% (22268/22268), 12.47 MiB | 18.72 MiB/s, done.
Resolving deltas: 100% (15150/15150), done.


In [14]:

!curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y

import os
os.environ['PATH'] += ":/root/.cargo/bin"

!cargo --version

info: downloading installer
warn: It looks like you have an existing rustup settings file at:
warn: /root/.rustup/settings.toml
warn: Rustup will install the default toolchain as specified in the settings file,
warn: instead of the one inferred from the default host triple.
info: profile set to 'default'
info: default host triple is x86_64-unknown-linux-gnu
warn: Updating existing toolchain, profile choice will be ignored
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: default toolchain set to 'stable-x86_64-unknown-linux-gnu'

  stable-x86_64-unknown-linux-gnu unchanged - rustc 1.86.0 (05f9846f8 2025-03-31)


Rust is installed now. Great!

To get started you may need to restart your current shell.
This would reload your PATH environment variable to include
Cargo's bin directory ($HOME/.cargo/bin).

To configure your current shell, you need to source
the corresponding env file under $HOME/.cargo.

This is usually done by running one of the following (note the 

In [3]:
%cd /content/candle/candle-examples/examples
!cargo run --example phi --release -- --model 2 \
  --prompt "A skier slides down a frictionless slope of height 40m and length 80m. What's the skier speed at the bottom?"



/content/candle/candle-examples/examples
    Updating crates.io index
     Locking 645 packages to latest compatible versions
      Adding anstyle-wincon v3.0.6 (available: v3.0.7)
      Adding fancy-regex v0.13.0 (available: v0.14.0)
      Adding gemm v0.17.1 (available: v0.18.2)
      Adding getrandom v0.2.16 (available: v0.3.2)
      Adding imageproc v0.24.0 (available: v0.25.0)
      Adding metal v0.27.0 (available: v0.32.0)
      Adding once_cell v1.19.0 (available: v1.21.3)
      Adding parquet v51.0.0 (available: v55.0.0)
      Adding pyo3 v0.22.6 (available: v0.24.2)
      Adding pyo3-build-config v0.22.6 (available: v0.24.2)
      Adding regex v1.10.6 (available: v1.11.1)
      Adding rubato v0.15.0 (available: v0.16.2)
      Adding safetensors v0.4.5 (available: v0.5.3)
      Adding thiserror v1.0.69 (available: v2.0.12)
      Adding wasm-bindgen-futures v0.4.45 (available: v0.4.50)
      Adding wasm-bindgen-test v0.3.45 (available: v0.3.50)
      Adding wasm-bindgen-test-mac

https://github.com/huggingface/candle/tree/main/candle-examples/examples/phi

https://github.com/ToluClassics/candle-tutorial

In [5]:
%cd /content/candle/candle-examples/examples/mistral
!cargo run --example mistral --release  -- --prompt 'Write helloworld code in Rust' --sample-len 150

/content/candle/candle-examples/examples/mistral
   Compiling candle-examples v0.9.1 (/content/candle/candle-examples)
    Finished `release` profile [optimized] target(s) in 36.17s
     Running `/content/candle/target/release/examples/mistral --prompt 'Write helloworld code in Rust' --sample-len 150`
avx: true, neon: false, simd128: false, f16c: true
temp: 0.00 repeat-penalty: 1.10 repeat-last-n: 64
Error: request error: https://huggingface.co/mistralai/Mistral-7B-v0.1/resolve/main/tokenizer.json: status code 401

Caused by:
    https://huggingface.co/mistralai/Mistral-7B-v0.1/resolve/main/tokenizer.json: status code 401


In [6]:
import os
os.environ['HUGGING_FACE_HUB_TOKEN'] = 'XXXX'

In [2]:
import os
os.environ['PATH'] += ":/root/.cargo/bin"

!cargo --version

cargo 1.86.0 (adf9b6ad1 2025-02-28)


In [4]:
%cd /content/candle/candle-examples/examples/mistral
!cargo run --example mistral --release  -- --prompt 'Write helloworld code in Rust'  --use-f32

/content/candle/candle-examples/examples/mistral
    Finished `release` profile [optimized] target(s) in 0.42s
     Running `/content/candle/target/release/examples/mistral --prompt 'Write helloworld code in Rust'`
avx: true, neon: false, simd128: false, f16c: true
temp: 0.00 repeat-penalty: 1.10 repeat-last-n: 64
Error: request error: https://huggingface.co/mistralai/Mistral-7B-v0.1/resolve/main/tokenizer.json: status code 401

Caused by:
    https://huggingface.co/mistralai/Mistral-7B-v0.1/resolve/main/tokenizer.json: status code 401


In [5]:
%cd /content/candle/candle-examples/examples/mistral
!cargo run --example mixtral --release  -- --prompt 'Write helloworld code in Rust'

/content/candle/candle-examples/examples/mistral
   Compiling candle-examples v0.9.1 (/content/candle/candle-examples)
    Finished `release` profile [optimized] target(s) in 24.90s
     Running `/content/candle/target/release/examples/mixtral --prompt 'Write helloworld code in Rust'`
avx: true, neon: false, simd128: false, f16c: true
temp: 0.00 repeat-penalty: 1.10 repeat-last-n: 64
Error: request error: https://huggingface.co/mistralai/Mixtral-8x7B-v0.1/resolve/main/tokenizer.json: status code 401

Caused by:
    https://huggingface.co/mistralai/Mixtral-8x7B-v0.1/resolve/main/tokenizer.json: status code 401


In [9]:
%cd /content/candle/candle-examples/examples/mistral
!cargo run --example mistral -- --use-f32 --release  -- --prompt 'Write helloworld code in Rust' --use-f32

/content/candle/candle-examples/examples/mistral
   Compiling proc-macro2 v1.0.95
   Compiling unicode-ident v1.0.18
   Compiling libc v0.2.172
   Compiling quote v1.0.40
   Compiling cfg-if v1.0.0
   Compiling autocfg v1.4.0
   Compiling syn v2.0.101
   Compiling once_cell v1.19.0
   Compiling libm v0.2.14
   Compiling num-traits v0.2.19
   Compiling zerocopy v0.8.25
   Compiling memchr v2.7.4
   Compiling bytemuck_derive v1.9.3
   Compiling ppv-lite86 v0.2.21
   Compiling crossbeam-utils v0.8.21
   Compiling getrandom v0.3.2
   Compiling either v1.15.0
   Compiling paste v1.0.15
   Compiling bytemuck v1.23.0
   Compiling num-complex v0.4.6
   Compiling shlex v1.3.0
   Compiling crossbeam-epoch v0.9.18
   Compiling cc v1.2.21
   Compiling rayon-core v1.12.1
   Compiling rand_core v0.9.3
   Compiling crossbeam-deque v0.8.6
   Compiling synstructure v0.13.2
   Compiling rand_chacha v0.9.0
   Compiling reborrow v0.5.5
   Compiling rand v0.9.1
   Compiling zerofrom-derive v0.1.6
   Compil

In [10]:
%cd /content/candle/candle-examples/examples/mistral
!cargo run --example mistral --release  -- --prompt 'Write helloworld code in Rust' --use-f32

/content/candle/candle-examples/examples/mistral
    Finished `release` profile [optimized] target(s) in 0.62s
     Running `/content/candle/target/release/examples/mistral --prompt 'Write helloworld code in Rust' --use-f32`
error: unexpected argument '--use-f32' found

Usage: mistral --prompt <PROMPT>

For more information, try '--help'.


In [13]:
%cd /content/candle/candle-examples/examples/mistral
!cargo run --example mistral --release -- --prompt 'Write helloworld code in Rust'

/content/candle/candle-examples/examples/mistral
    Finished `release` profile [optimized] target(s) in 0.36s
     Running `/content/candle/target/release/examples/mistral --prompt 'Write helloworld code in Rust'`
avx: true, neon: false, simd128: false, f16c: true
temp: 0.00 repeat-penalty: 1.10 repeat-last-n: 64
Error: request error: https://huggingface.co/mistralai/Mistral-7B-v0.1/resolve/main/tokenizer.json: status code 401

Caused by:
    https://huggingface.co/mistralai/Mistral-7B-v0.1/resolve/main/tokenizer.json: status code 401


In [ ]:
toml
[[example]]
name = "mistral"
path = "examples/mistral/main.rs"

In [ ]:
toml
[[example]]
name = "mistral"
path = "examples/mistral/main.rs"
required-features = ["f32"]  # Add this line

In [15]:
%cd /content/candle/candle-examples/examples/mistral
!cargo run --example mistral --release -- --prompt 'Write helloworld code in Rust' --features="f32"

/content/candle/candle-examples/examples/mistral
error: target `mistral` in package `candle-examples` requires the features: `f32`
Consider enabling them by passing, e.g., `--features="f32"`


In [16]:
%cd /content/candle/candle-examples/examples/mistral
!cargo run --example mistral --release --features="f32" -- --prompt 'Write helloworld code in Rust'

/content/candle/candle-examples/examples/mistral
error: the package 'candle-examples' does not contain this feature: f32
help: there is a similarly named feature: mkl


In [ ]:
/content/candle/Cargo.toml

In [ ]:
toml
 [features]
 default = ["cuda", "cublas-lt"]
 f32 = []

In [17]:
%cd /content/candle
!cargo build --release --features f32

/content/candle
error: failed to parse manifest at `/content/candle/Cargo.toml`

Caused by:
  this virtual manifest specifies a `features` section, which is not allowed


In [18]:
%cd /content/candle/candle-examples/examples/mistral
!cargo run --example mistral --release --features f32 -- --prompt 'Write helloworld code in Rust'

/content/candle/candle-examples/examples/mistral
error: failed to parse manifest at `/content/candle/candle-examples/Cargo.toml`

Caused by:
  error inheriting `edition` from workspace root manifest's `workspace.package.edition`

Caused by:
  failed to parse manifest at `/content/candle/Cargo.toml`

Caused by:
  this virtual manifest specifies a `features` section, which is not allowed


In [22]:
%cd /content/candle
!cargo build

/content/candle
error: failed to parse manifest at `/content/candle/Cargo.toml`

Caused by:
  this virtual manifest specifies a `features` section, which is not allowed


In [2]:
import os
os.environ['PATH'] += ":/root/.cargo/bin"

!cargo --version

cargo 1.86.0 (adf9b6ad1 2025-02-28)


In [11]:
!rm -rf /content/candle

In [7]:
%cd /content/candle/candle-examples/examples


!cargo run --example falcon --release -- --prompt "Flying monkeys are"

/content/candle/candle-examples/examples
error: failed to parse manifest at `/content/candle/candle-examples/Cargo.toml`

Caused by:
  error inheriting `edition` from workspace root manifest's `workspace.package.edition`

Caused by:
  failed to parse manifest at `/content/candle/Cargo.toml`

Caused by:
  this virtual manifest specifies a `features` section, which is not allowed


In [9]:
%cd /content/candle/candle-examples/examples
!cargo run --example qwen --release  -- --prompt "Hello there "

/content/candle/candle-examples/examples
error: failed to parse manifest at `/content/candle/candle-examples/Cargo.toml`

Caused by:
  error inheriting `edition` from workspace root manifest's `workspace.package.edition`

Caused by:
  failed to parse manifest at `/content/candle/Cargo.toml`

Caused by:
  this virtual manifest specifies a `features` section, which is not allowed


In [10]:
%cd /content/candle/candle-examples/examples
!cargo run --example phi --release -- --model 2 \
  --prompt "A skier slides down a frictionless slope of height 40m and length 80m. What's the skier speed at the bottom?"



/content/candle/candle-examples/examples
error: failed to parse manifest at `/content/candle/candle-examples/Cargo.toml`

Caused by:
  error inheriting `edition` from workspace root manifest's `workspace.package.edition`

Caused by:
  failed to parse manifest at `/content/candle/Cargo.toml`

Caused by:
  this virtual manifest specifies a `features` section, which is not allowed


In [15]:
%cd /content/candle/candle-examples/examples
!cargo run --example qwen --release  -- --prompt "Hello there "

/content/candle/candle-examples/examples
    Updating crates.io index
     Locking 645 packages to latest compatible versions
      Adding anstyle-wincon v3.0.6 (available: v3.0.7)
      Adding fancy-regex v0.13.0 (available: v0.14.0)
      Adding gemm v0.17.1 (available: v0.18.2)
      Adding getrandom v0.2.16 (available: v0.3.2)
      Adding imageproc v0.24.0 (available: v0.25.0)
      Adding metal v0.27.0 (available: v0.32.0)
      Adding once_cell v1.19.0 (available: v1.21.3)
      Adding parquet v51.0.0 (available: v55.0.0)
      Adding pyo3 v0.22.6 (available: v0.24.2)
      Adding pyo3-build-config v0.22.6 (available: v0.24.2)
      Adding regex v1.10.6 (available: v1.11.1)
      Adding rubato v0.15.0 (available: v0.16.2)
      Adding safetensors v0.4.5 (available: v0.5.3)
      Adding thiserror v1.0.69 (available: v2.0.12)
      Adding wasm-bindgen-futures v0.4.45 (available: v0.4.50)
      Adding wasm-bindgen-test v0.3.45 (available: v0.3.50)
      Adding wasm-bindgen-test-mac

In [ ]:
/content/

In [18]:
%cd /content/candle/candle-examples/examples
!cargo run --example onnx --features=onnx --release -- --image /content/candle/candle-examples/examples/yolo-v8/assets/bike.jpg

/content/candle/candle-examples/examples
    Finished `release` profile [optimized] target(s) in 0.37s
     Running `/content/candle/target/release/examples/onnx --image /content/candle/candle-examples/examples/yolo-v8/assets/bike.jpg`
loaded image Tensor[dims 3, 224, 224; f32]
squeezenet1.1-7.onnx [00:00:00] [███████████] 4.73 MiB/4.73 MiB 23.01 MiB/s (0s)unicycle, monocycle                               : 83.27%
ballplayer, baseball player                       : 3.65%
bearskin, busby, shako                            : 1.60%
military uniform                                  : 0.76%
trombone                                          : 0.75%


In [ ]:
/content/candle/candle-examples/examples/yolo-v8/assets/bike.jpg

In [ ]:
%cd /content/candle/candle-examples/examples
!cargo run --example phi --release -- --model 2 \
  --prompt "A skier slides down a frictionless slope of height 40m and length 80m. What's the skier speed at the bottom?"

/content/candle/candle-transformers/tests/generation_tests.rs

In [20]:
%cd /content/candle/candle-transformers
!cargo run -- --tests generation_tests --release -- --model 2 \
  --prompt "A skier slides down a frictionless slope of height 40m and length 80m. What's the skier speed at the bottom?"


/content/candle/candle-transformers
error: a bin target must be available for `cargo run`


In [ ]:
%cd /content/candle/candle-transformers
!cargo test --test generation_tests --release -- --model 2 --prompt "A skier slides down a frictionless slope of height 40m and length 80m. What's the skier speed at the bottom?"

In [21]:
!cargo test --test generation_tests --release -- --model 2 --prompt "A skier slides down a frictionless slope of height 40m and length 80m. What's the skier speed at the bottom?"

   Compiling bytemuck v1.23.0
   Compiling either v1.15.0
   Compiling rayon v1.10.0
   Compiling num-complex v0.4.6
   Compiling half v2.6.0
   Compiling pulp v0.21.5
   Compiling dyn-stack v0.13.0
   Compiling pulp v0.18.22
   Compiling gemm-common v0.18.2
   Compiling dyn-stack v0.10.0
   Compiling gemm-common v0.17.1
   Compiling memchr v2.7.4
   Compiling gemm-f32 v0.18.2
   Compiling serde_json v1.0.140
   Compiling gemm-f16 v0.18.2
   Compiling gemm-f32 v0.17.1
   Compiling gemm-c32 v0.18.2
   Compiling gemm-c64 v0.18.2
   Compiling gemm-f64 v0.18.2
   Compiling tracing-core v0.1.33
   Compiling tracing v0.1.41
   Compiling gemm v0.18.2
   Compiling gemm-f16 v0.17.1
   Compiling num v0.4.3
   Compiling safetensors v0.4.5
   Compiling gemm-c32 v0.17.1
   Compiling gemm-f64 v0.17.1
   Compiling gemm-c64 v0.17.1
   Compiling ug v0.4.0
   Compiling gemm v0.17.1
   Compiling aho-corasick v1.1.3
   Compiling regex-syntax v0.8.5
   Compiling candle-core v0.9.1 (/content/candle/candle-c

In [22]:
%cd /content/candle/candle-transformers
!cargo test --test generation_tests --release

/content/candle/candle-transformers
    Finished `release` profile [optimized] target(s) in 0.65s
     Running tests/generation_tests.rs (/content/candle/target/release/deps/generation_tests-bbed9e1180fd543a)

running 5 tests
test sample_with_temperature ... ok
test sample_with_top_k ... ok
test sample_with_top_p ... ok
test sample_with_zero_temperature ... ok
test sample_gumbel ... ok

test result: ok. 5 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.40s



الناتج الذي ظهر لك يعني أنك قمت بتشغيل ملف الاختبارات generation_tests.rs بنجاح بعد اتباع الخطوات التي تم ذكرها سابقاً.

دعنا نشرح كل سطر من الناتج:

Finished release profile [optimized] target(s) in 0.65s:

هذا السطر يخبرك أن عملية بناء ملفات الكود اللازمة للتشغيل ("release profile") تمت بنجاح في 0.65 ثانية.
كلمة "optimized" تعني أن الملفات تم بناؤها بأفضل أداء ممكن.
Running tests/generation_tests.rs (...):

هذا السطر يوضح لك مسار ملف الاختبارات الذي تم تشغيله.
running 5 tests:

هذا السطر يخبرك أن ملف الاختبارات generation_tests.rs يحتوي على 5 اختبارات مختلفة، وتم تشغيلهم جميعاً.
test sample_with_temperature ... ok:

كل سطر يبدأ ب "test" يمثل اختبار محدد ضمن الملف.
كلمة "ok" تعني أن الاختبار تم اجتيازه بنجاح.
test result: ok. 5 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.40s:

هذا السطر هو ملخص لنتائج تشغيل جميع الاختبارات.
5 passed: يعني أن جميع الاختبارات الخمسة تم اجتيازها بنجاح.
0 failed: يعني أنه لم يحدث أي فشل في أي من الاختبارات.
0 ignored: يعني أنه لم يتم تجاهل أي من الاختبارات.
finished in 0.40s: يعني أن جميع الاختبارات تم تشغيلها في 0.40 ثانية.
بإختصار:

تم بناء الملفات اللازمة للتشغيل، ثم تم تشغيل جميع الاختبارات الخمسة بنجاح وبدون أي مشاكل في أقل من ثانية. هذا يؤكد أن ملف الاختبارات generation_tests.rs يعمل بشكل سليم.

Rate this answer
